azure_data_cosmos_engine/src/query/pipeline.rs

```diff
@@ -295,14 +297,129 @@ impl QueryPipeline {
 
         Ok(PipelineResponse {
             items,
             requests,
             terminated: self.terminated,
         })
     }
+
+    /// Creates a new read many pipeline.
+    ///
+    /// # Parameters
+    /// * `item_identities` - An iterator that produces the [`ItemIdentity`]s to be read.
+    /// * `pkranges` - An iterator that produces the [`PartitionKeyRange`]s for the container to use for query generation.
+    /// * `pk_kind` - The partition key kind.
+    /// * `pk_version` - The partition key version.
+    #[tracing::instrument(level = "debug", skip_all, err)]
+    pub fn for_read_many(
+        item_identities: impl IntoIterator<Item = ItemIdentity>,
+        pkranges: impl IntoIterator<Item = PartitionKeyRange>,
+        pk_kind: &str,
+        pk_version: u32,
+    ) -> crate::Result<Self> {
+        let mut pkranges: Vec<PartitionKeyRange> = pkranges.into_iter().collect();
+        tracing::debug!(?pkranges, "creating readmany pipeline2");
+
+        // Grab item identities and start grouping them by partition key range.
+        // Output should be a list of tuples of (pkrangeid, query_string) to go over and generate item producers for.
+        let item_identities: Vec<ItemIdentity> = item_identities.into_iter().collect();
+        tracing::debug!(?item_identities, "received item identities for read many");
+        // Group items by their partition key range ID.
+        let items_by_range =
+            Self::partition_items_by_range(item_identities, &mut pkranges, pk_kind, pk_version);
+        tracing::debug!(
+            ?items_by_range,
+            "grouped item identities by partition key range"
+        );
+        // Create query chunks from the partitioned items, splitting total list into 1000 max item queries.
+        // Each chunk is represented as vector of mappings of partition key range IDs to lists of tuples containing the original index, item ID, and partition key value.
+        let query_chunks = Self::create_query_chunks_from_partitioned_items(&items_by_range);
+        tracing::debug!(?query_chunks, "created query chunks from partitioned items");
+
+        // Create the item producer for read many.
+        let producer = ItemProducer::read_many(query_chunks);
+
+        // We are building the pipeline outside-in.
+        // That means the first node we push will be the first node executed.
+        // This is relevant for nodes like OFFSET and LIMIT, which need to be ordered carefully.
+        let pipeline: Vec<Box<dyn PipelineNode>> = Vec::new();
+
+        Ok(Self {
+            query: String::new(),
+            pipeline,
+            producer,
+            terminated: false,
+        })
+    }
+
+    // Groups items by their partition key range ID efficiently while preserving original order.
+    // Returns a mapping of partition key range IDs to lists of tuples containing the original index, item ID, and partition key value.
+    fn partition_items_by_range(
```

Comment:
nit: In Rust you should generally only put a function in an `impl` block if:

* It accepts `Self` (or `&Self`, `&mut Self`, etc) as a parameter, and/or
* It returns `Self` (or `Result<Self>`, `Option<Self>`, etc.)

This should just be a bare `fn` outside the `impl` block. It's quite idiomatic to have `fn`s directly inside a module, rather than grouping them into types (like static classes in .NET/Java/etc.)